### PR TITLE
added note by commitment request handler

### PR DESCRIPTION
**Pull request by Valentine1898**

### CODE REVIEW

File: packages/router/src/grpc/view-protocol-server/note-by-commitment.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+};
+
+export const handleNoteByCommitmentReq = async (
+  req: NoteByCommitmentRequest,
+  services: ServicesInterface,
+): Promise<NoteByCommitmentResponse> => {
+  const { indexedDb } = await services.getWalletServices();
+  if (!req.noteCommitment) return new NoteByCommitmentResponse();
+
+  // TODO  req.await_detection processing
```
Comment (grod220) [inline packages/router/src/grpc/view-protocol-server/note-by-commitment.ts:19]: We have an issue for await detection somewhere. Can you add a todo in that issue to implement this?

File: packages/router/src/grpc/view-protocol-server/note-by-commitment.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+export const isNoteByCommitmentRequest = (req: ViewReqMessage): req is NoteByCommitmentRequest => {
+  return req.getType().typeName === NoteByCommitmentRequest.typeName;
+};
+
+export const handleNoteByCommitmentReq = async (
+  req: NoteByCommitmentRequest,
+  services: ServicesInterface,
+): Promise<NoteByCommitmentResponse> => {
+  const { indexedDb } = await services.getWalletServices();
+  if (!req.noteCommitment) return new NoteByCommitmentResponse();
```
Comment (grod220) [inline packages/router/src/grpc/view-protocol-server/note-by-commitment.ts:17]: I feel like we should throw an error. Or does the Rust implementation send an empty response like this?

File: packages/storage/src/indexed-db/indexed-db.test.ts [hunk 3]
```diff
@@ -191,6 +191,15 @@ describe('IndexedDb', () => {
...
     });
+
+    it('should be able to set/get by commitment', async () => {
+      const db = await IndexedDb.initialize({ ...generateInitialProps() });
+
+      await db.saveSpendableNote(newNote);
+      const noteByCommitment = await db.getNoteByCommitment(newNote.noteCommitment!);
+
+      expect(newNote.equals(noteByCommitment)).toBeTruthy();
+    });
```
Comment (grod220) [inline packages/storage/src/indexed-db/indexed-db.test.ts:202]: Can you put an undefined case as well?

File: packages/router/src/grpc/view-protocol-server/note-by-commitment.ts (hunk 2 above)
Comment (Valentine1898) [reply packages/router/src/grpc/view-protocol-server/note-by-commitment.ts:17]: Rust imlementation throws an error. I noticed that in one more request we return an empty response where we should throw an error
https://github.com/penumbra-zone/web/blob/89d4ec182ec5946410bea0b057e45afa00be6552/packages/router/src/grpc/view-protocol-server/tx-info-by-hash.ts#L18

File: packages/router/src/grpc/view-protocol-server/note-by-commitment.ts (hunk 1 above)
Comment (Valentine1898) [reply packages/router/src/grpc/view-protocol-server/note-by-commitment.ts:19]: Yes, I was planning to add a TODO link to https://github.com/penumbra-zone/web/issues/233 after the merge

**Review by grod220 — APPROVED** [review]

**Pull request merged**